In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

df = pd.read_csv('https://www.ishelp.info/data/insurance.csv')

# Generate dummy variables
for col in df:  
  if not pd.api.types.is_numeric_dtype(df[col]):
    df = pd.get_dummies(df, columns=[col], prefix=col, dtype=int)
    
df.head()

age     bmi  children      charges  sex_female  sex_male  smoker_no  \
0   19  27.900         0  16884.92400           1         0          0   
1   18  33.770         1   1725.55230           0         1          1   
2   28  33.000         3   4449.46200           0         1          1   
3   33  22.705         0  21984.47061           0         1          1   
4   32  28.880         0   3866.85520           0         1          1   

   smoker_yes  region_northeast  region_northwest  region_southeast  \
0           1                 0                 0                 0   
1           0                 0                 0                 1   
2           0                 0                 0                 1   
3           0                 0                 1                 0   
4           0                 0                 1                 0   

   region_southwest  
0                 1  
1                 0  
2                 0  
3                 0  
4                 0

In [2]:
y = df['bmi']
X = df.drop(columns=['bmi', 'sex_female', 'region_northwest', 'region_northeast', 'smoker_yes', 'children', "sex_male"]).assign(const=1)

# Run the multiple linear regression model
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())  # View results

                            OLS Regression Results                            
Dep. Variable:                    bmi   R-squared:                       0.181
Model:                            OLS   Adj. R-squared:                  0.178
Method:                 Least Squares   F-statistic:                     58.76
Date:                Tue, 28 Jan 2025   Prob (F-statistic):           2.25e-55
Time:                        08:28:20   Log-Likelihood:                -4183.8
No. Observations:                1338   AIC:                             8380.
Df Residuals:                    1332   BIC:                             8411.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
age                 -0.0289      0.013  

In [3]:
y = df['bmi']
X = df.drop(columns=['bmi', 'sex_female', 'region_northwest', 'region_northeast', 'smoker_yes', 'children']).assign(const=1)

# Run the multiple linear regression model
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())  # View results

                            OLS Regression Results                            
Dep. Variable:                    bmi   R-squared:                       0.183
Model:                            OLS   Adj. R-squared:                  0.179
Method:                 Least Squares   F-statistic:                     49.54
Date:                Tue, 28 Jan 2025   Prob (F-statistic):           4.10e-55
Time:                        08:28:20   Log-Likelihood:                -4182.3
No. Observations:                1338   AIC:                             8379.
Df Residuals:                    1331   BIC:                             8415.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
age                 -0.0284      0.013  

In [4]:
y = df['bmi']
X = df.drop(columns=['bmi', 'sex_female', 'region_northwest', 'region_northeast', 'region_southwest', 'smoker_yes', 'children', "sex_male"]).assign(const=1)

# Run the multiple linear regression model
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())  # View results

                            OLS Regression Results                            
Dep. Variable:                    bmi   R-squared:                       0.171
Model:                            OLS   Adj. R-squared:                  0.169
Method:                 Least Squares   F-statistic:                     68.84
Date:                Tue, 28 Jan 2025   Prob (F-statistic):           5.10e-53
Time:                        08:28:20   Log-Likelihood:                -4191.5
No. Observations:                1338   AIC:                             8393.
Df Residuals:                    1333   BIC:                             8419.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
age                 -0.0281      0.013  

In [5]:
y = df['bmi']
X = df.drop(columns=['bmi', 'sex_female', 'region_northwest', 'region_southwest', 'region_northeast', 'smoker_yes', 'children', 'smoker_no', 'sex_male', 'region_southeast']).assign(const=1)

# Run the multiple linear regression model
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())  # View results

                            OLS Regression Results                            
Dep. Variable:                    bmi   R-squared:                       0.042
Model:                            OLS   Adj. R-squared:                  0.041
Method:                 Least Squares   F-statistic:                     29.32
Date:                Tue, 28 Jan 2025   Prob (F-statistic):           3.44e-13
Time:                        08:28:20   Log-Likelihood:                -4288.4
No. Observations:                1338   AIC:                             8583.
Df Residuals:                    1335   BIC:                             8598.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
age            0.0238      0.012      1.955      0.0

Final BMI Model is the first model

The first model included all variables, resulting in an adjusted R-squared of 0.178 and an F-statistic of 58.76, with all variables having p-values less than 0.05, indicating statistical significance. The "age" coefficient was negative, suggesting that as age increases, BMI decreases. The "charges" variable had a small positive effect, showing a slight increase in BMI with rising charges. The southeast and southwest regions showed significant positive effects on BMI compared to the reference group.

In the new second model, we added the "sex_male" variable, resulting in an adjusted R-squared of 0.179 and an F-statistic of 49.54. While the adjusted R-squared remained similar, the F-statistic decreased, indicating slightly less overall explanatory power. The coefficient for "sex_male" was positive, with a p-value of 0.082, suggesting a marginal effect on BMI, but it was close to being statistically significant. The "age" and "charges" coefficients remained stable, and the southeast region continued to have a positive effect on BMI.

The third model, dropping additional variables ("region_southwest" and "smoker_yes"), resulted in an adjusted R-squared of 0.169. Although the adjusted R-squared decreased slightly, the F-statistic increased to 68.84, indicating that the model still explained a slightly larger proportion of variance despite the reduced number of predictors. The coefficients for "age" and "charges" remained similar, and the southeast region continued to show a positive effect on BMI.

In the fourth model, excluding most categorical variables, including "region_southeast," the explanatory power dropped significantly, with an adjusted R-squared of 0.041. While "charges" remained significant, the overall model lost explanatory power, highlighting the importance of including key categorical variables.

Given the results from all four models, the first model remains the best choice. It strikes a good balance between explanatory power (adjusted R-squared of 0.178) and statistical significance, with all predictors being relevant to the model. The second model’s inclusion of "sex_male" didn’t improve the model enough to offset the slight drop in explanatory power. Meanwhile, the third and fourth models, while showing interesting trends with their respective changes in variables, did not provide the same robustness or explanatory power as the first model. Therefore, the first model was chosen as the final model due to its comprehensive explanatory power and statistical significance.

In [6]:
y = df['age']
X = df.drop(columns=['age', 'sex_female', "smoker_no", "region_northeast", "region_northwest", "region_southwest", 'region_southeast', "sex_male", "children"]).assign(const=1)

# Run the multiple linear regression model
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())  # View results

                            OLS Regression Results                            
Dep. Variable:                    age   R-squared:                       0.271
Model:                            OLS   Adj. R-squared:                  0.269
Method:                 Least Squares   F-statistic:                     165.3
Date:                Tue, 28 Jan 2025   Prob (F-statistic):           4.01e-91
Time:                        08:28:20   Log-Likelihood:                -5222.4
No. Observations:                1338   AIC:                         1.045e+04
Df Residuals:                    1334   BIC:                         1.047e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
bmi           -0.1393      0.057     -2.454      0.0

This model looks at how age is related to BMI, charges, and smoking status. The R-squared value of 0.271 shows that about 27% of the variation in age can be explained by these factors, with a very significant F-statistic of 165.3 (p-value 4.01e-91). The BMI coefficient of -0.1393 means that for each unit increase in BMI, age decreases by 0.1393 years. The charges coefficient of 0.0010 suggests that as charges increase, age goes up by 0.001 years. The "smoker_yes" variable has a coefficient of -24.6796, which implies smokers are, on average, about 24.68 years younger than non-smokers (which seems a bit weird). Still, all variables are statistically significant, so the model does a decent job predicting age.

2) Residuals are basically the difference between the actual values and the predicted values from a model. They're kind of like the "errors" or the parts the model couldn't explain. For instance, in a model predicting age based on BMI, charges, and smoking status, the residual would be the difference between a person’s actual age and the age the model predicted. If the residual is small, that means the model did a good job. If it’s large, the model might not be capturing something important. Checking residuals helps us see how well the model fits the data and if there are any patterns it’s missing.